src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+  plus (Const b1) (Const b2) = Const $ b1 `plus` b2
+  plus (Max uses1) (Max uses2) = Max uses' where
+    uses' = concat $ flip map uses1 (\use1 ->
+      flip map uses2 \use2 -> plus use1 use2)
+      -- TODO: Detect overlapping uses here?
+  plus u1@(Max _) use2
+    | use2 == zero = u1
+    | otherwise = plus u1 (Max [use2])
+  plus use1 u2@(Max _)
+    | use1 == zero = u2
+    | otherwise = plus (Max [use1]) u2
+  plus (Sum uses1) (Sum uses2) = Sum $ uses1 ++ uses2
+  plus (Sum uses1) use2
+    -- TODO Is there a way to deduplicate these zero checks?
+    | use2 == zero = Sum uses1
+    | otherwise = Sum $ uses1 ++ [use2]
+  plus use1 (Sum uses2)
+    | use1 == zero = Sum uses2
+    | otherwise = Sum $ use1:uses2
+  plus use1 use2
+    | use1 == zero = use2
+    | use2 == zero = use1
+    | otherwise = Sum [use1, use2]
+
+-- === Interpreting Access into Use ===
+
+-- The main interpreter is pretty straightforward.  The main trick is
+-- just semantics: an indexing expression becomes a (nested) `At`,
+-- which represents a data-space loop that checks that the data-space
+-- variable equals the indexing expression.
+interp :: Access iter -> Use iter
+interp (Location [] ct) = Const ct
+interp (Location (expr:exprs) ct) = At expr rest where
+  rest = interp $ Location exprs ct
+interp (Any accesses) = foldl max zero uses where
+  uses = map interp accesses
+interp (All accesses) = foldl plus zero uses where
+  uses = map interp accesses
+interp (ForEach b access) = iterate False b $ interp access
+
+-- The interesting function is resolution of iteration-space loops.
+-- We have to construct a Use that can refer to subsets of data space
+-- (as opposed to single points, as in Access).
+-- The `Bool` argument means "have we already observed that upstream
+-- indexing touches different parts of the object?"  Even if the
+-- answer is "yes", we do still have to recur, because we need to
+-- translate the interation-space binder into data space.
+
+iterate :: Bool -> (NameBinder AtomNameC iter iter')
+          -> Use iter' -> (Use iter)
+iterate True _ (Const ct) = Const ct
+iterate False _ (Const ct) =
+  -- If we haven't proven upstream that indexing is injective,
+  -- we have to assume here that it collides with itself.
+  -- e.g., for i. xs
+  -- Note: If we have a bound on the size of the index set traversed
+  -- by `ib`, then we can compute a bound on the number of accesses,
+  -- but there's no need to do that until we make inlining decisions
+  -- on a more fine-grained basis than "was this used more than once?"
+  if ct == zero then zero else Const Unbounded
+iterate injective ib (At expr use) = At expr' use' where
+  (expr', injective') =
+    flip runState injective $ iterateIxExpr ib expr
+  use' = iterate injective' ib use
+iterate injective ib (Max uses) =
+  foldl max zero $ map (iterate injective ib) uses
+iterate injective ib (Sum uses) =
+  foldl plus zero $ map (iterate injective ib) uses
```

Comment:
This rule essentially reassociates the sum coming from the iteration over the sum in the use, right?

The previous one seems like it might be an underapproximation? `sum for i. (max xs.i) > max (sum for i. xs.i)` and we go from the first one to the second one

Reply:
Good one!  I'm not sure how to deal with this, because we really do need to push `iterate` as far down as we can, to get rid of the iteration-space binders.

For now I think it's actually fine, due to an accident of the upstream Access structure; left a detailed comment in the code.

Reply:
Yeah I think I agree. Consider `sum for i. (max x y)` (`x` and `y` are uses that might contain `i`). My take on this would be that for this to be an underapproximation, we'd need `x` to be sometimes larger than `y` and sometimes smaller than `y`. But I think that `Use` expressions cannot be monotonic in this way, at least not given the current set of constructors? You understand this better, so could you perhaps finish this argument? I'm pretty sure it will be much simpler than whatever is in the code right now.

Reply:
I don't follow your counterargument. Think of `x` and `y` as a count of uses in the two case branches. If one branch is consistently larger than the other then commuting a max with a sum is actually sound and will return in the exact same value being returned.